src/pybind/mgr/ansible/ansible_runner_svc.py

```diff
@@ -393,3 +394,119 @@ def get_hosts_in_group(self, group):
                                             " inventory".format(group))
 
         return json.loads(response.text)["data"]["members"]
+
+
+class InventoryGroup(collections.MutableSet):
+    """ Manages an Ansible Inventory Group
+    """
+    def __init__(self, group_name, ars_client):
+        """Init the group_name attribute and
```

Comment:
Looks, great! The only issue I see here is that as the data is fetched only during object initialization, it could be soon (and inadvertently) outdated, and besides, multiple successive invocations of this method would cause HTTP requests. I'd suggest lazy-loading approach + flyweight pattern + caching:

You may start using a ['flyweight' initialization](https://www.tutorialspoint.com/python_design_patterns/python_design_patterns_flyweight.htm) (you only create one `InventoryGroup` per `group_name`, and then you use caching for requests:
```py
class InventoryGroup(collections.MutableSet):
  registry = {}
  MAX_TTL = 5 # seconds

  # flyweight pattern: multiple InventoryGroup() instantiations for same group return the same object
  def __new__(cls, group_name, ars_client):
    try:
      ig = cls.registry[group_name]
    except KeyError:
       # this calls __init__ too
       ig = super(InventoryGroup, cls).__new__(group_name, ars_client)
       cls.registry[group_name] = ig
    return ig

  def __init__(self, group_name, ars_client):
     # no data fetching, quick initialization
     self.group_name = group_name
     self.ars_client = ars_client
     self.url_group = URL_MANAGE_GROUP.format(group_name=self.group_name)
     # this could be done in a list or a class
     self.__element_cache_item = None
     self.__element_cache_ts = 0
```

And then, for the methods that actually need fresh data you only need to create a TTL-cached property:
```py
  # we create a property instead of a plain class variable
  # if I had submitted the @ttl_cache decorator I've been working on this could be more easily done
  @property
  def elements(self):
    if time.time() - self.__element_cache_ts > self.MAX_TTL:
      response = self.ars_client.http_get(self.url_group)
        if response.status_code == 404:
          raise WhateverException
	self.__element_cache_item = set(json.loads(response.text)["data"]["members"])
    return self.__element_cache_item

    def __contains__(self, host):
        """ Check if the host is in the group
         : host: Check if hosts is in Ansible Inventory Group
        """
        # No need to modify anything, it use the latest value if it's still fresh, or it'll refresh it if needed
        return host in  self.elements
```

This way you ensure objects and data within are always keep fresh (with a `MAX_TTL` delay).

Reply:
As is usual, this is an incredible and awesome suggestion. Thanks Ernesto!.
In this case i preferred not to follow it because the main use of this type of objects make very improbable the repetitive invocation of methods, so the "cache" utilization will be very limited, mostly never used.